### DISCUSSION TITLE
Struggling to get custom fonts working

### DISCUSSION THREAD

**Discussion by FraserElectronics** [AvaloniaUI/Avalonia#14298] [Q&A] [answered]
I am trying to include a custom font in my Avalonia project using the instructions here: [https://docs.avaloniaui.net/docs/guides/styles-and-resources/how-to-use-fonts#:~:text=In%20your%20Avalonia%20application%2C%20open,reference%20it%20in%20your%20application.](url)

In /Assets/Fonts I have Montserrat.ttf. 
In app.axaml I have:

```
<Application.Resources>
	<FontFamily x:Key="Montserrat">avares://Assets/Fonts#Montserrat</FontFamily>
</Application.Resources>
```

I then try and use the font like so:

```
<TextBlock Text="Some test text" FontFamily="{StaticResource Montserrat}" FontSize="32" Foreground="White"/> 
```

However, I get the error: 

```
Could not create glyphTypeface.
```

If I refer to the font directly with:

```
<TextBlock Text="Some test text" FontFamily="/Assets/Fonts/Montserrat.ttf#Montserrat" FontSize="32" Foreground="White"/>
```

It works fine.
The font is set as an AvaloniaResource.

What am I doing wrong?

Andy

**Comment by timunie** [comment, marked as answer]
I believe you need to add your project name here:

```diff
<Application.Resources>
!	<FontFamily x:Key="Montserrat">avares://[MyProject]/Assets/Fonts#Montserrat</FontFamily>
</Application.Resources>

**Reply by FraserElectronics** [reply]
@timunie 
Tim, many thanks for that 👍
Looking back at the documentation it does actually show the project name which I had managed to miss 😒

Thanks for your help.

Andy

**Reply by deallocator** [reply]
And I stumbled upon the fact that you should never use '-' or '_' in a project name.

**Reply by princefishthrower** [reply]
I guess I don't understand what the `Fonts#<FontName>` means. If I have for example a single .ttf in `/Assets/Fonts/`, i.e. `/Assets/Fonts/MyFont.ttf`, am I supposed to use `/Assets/Fonts#MyFont`? Also the `avares://` is unclear to me... I don't need or use it anywhere else in my app for resources...

**Reply by stevemonaco** [reply]
> If I have for example a single .ttf in `/Assets/Fonts/`, i.e. `/Assets/Fonts/MyFont.ttf`, am I supposed to use `/Assets/Fonts#MyFont`? Also the `avares://` is unclear to me... I don't need or use it anywhere else in my app for resources..

Fonts are a bit weird here because there are various sources for FontCollections, including coming from Assets, System fonts, or potentially automatically loaded by Avalonia (Infer). I typically use the full reference here up to the folder that contains the font. eg. `avares://MyProject/Assets/Fonts#'. I assume the design helps font fallbacks be a bit smarter when the exact font isn't available.

What comes after the '#' is more complicated. It's not based on the filename, but rather it's the OTF Typography Name (ID 16, formerly Preferred Family) or, if empty, then the OTF Family Name (ID 1). You can [read the format docs here](https://learn.microsoft.com/en-us/typography/opentype/spec/name#name-ids).

Now, how do you actually know what this is? I use [fontdrop](https://fontdrop.info/), drop the font, and check the Data tab. From there, you can find the `preferredFamily` (Typography) or `fontFamily`. Remove any language prefixes (eg. `en ` in `en Font Awesome 6 Free`), then put that after the '#'.

**Reply by princefishthrower** [reply]
Ahhhh so this comes down to the actual font name within the data of the file itself. I can't just rename my .ttf file and use that.

Anyway, thank you for the excellent clarifications!
